### PR TITLE
Bulk array validation

### PR DESCRIPTION
**Pull request by tlively**
Add validation rules (prose and formalism) for `array.fill`, `array.copy`,
`array.init_data`, and `array.init_elem`. As part of the rules for `array.copy`,
extend the `matches` relation to storage types.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #401** <a href="https://app.graphite.dev/github/pr/WebAssembly/gc/401" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/gc/401?utm_source=stack-comment).

**Review by rossberg — APPROVED** [review]
LGTM modulo comments.

File: document/core/valid/matching.rst [hunk 1]
```diff
@@ -147,9 +147,6 @@ A :ref:`value type <syntax-valtype>` :math:`\valtype_1` matches a :ref:`value ty
    }
 
 
-.. index:: result type, value type
-.. _match-resulttype:
-
```
Comment (rossberg) [inline document/core/valid/matching.rst:152]: Did you want to delete those?

File: document/core/valid/instructions.rst [hunk 2]
```diff
@@ -598,6 +598,131 @@ Aggregate Reference Instructions
...
+     \qquad
+     C.\CDATAS[y] = {\ok}
+   }{
+     C \vdashinstr \ARRAYINITDATA~x~y : [(\REF~\NULL~x)~\I32~\I32~\I32] \to []
+   }
+
+
+.. _valid-array.init_elem:
+
+:math:`\ARRAYINITELEM~x~y
```
Comment (rossberg) [inline document/core/valid/instructions.rst:696]: This seems to be missing a closing tick.

File: document/core/valid/matching.rst (hunk 1 above)
Comment (tlively) [reply document/core/valid/matching.rst:152]: nope, thanks

File: document/core/valid/matching.rst [hunk 3]
```diff
@@ -146,6 +146,23 @@ A :ref:`value type <syntax-valtype>` :math:`\valtype_1` matches a :ref:`value ty
      C \vdashvaltypematch \BOT \matchesvaltype \valtype
    }
 
+.. index:: storage type
+.. _match-storagetype:
+
+Storage Types
```
[the remark this answers is not in the document]
Comment (tlively) [reply document/core/valid/matching.rst]: I've now also removed the macros that will be added back in by #382.

**Comment by tlively** [comment]
We'll wait on #382 to land this, since it no longer builds independently.

**Pull request merged**